Content.Server/Construction/Completions/BuildMachine.cs

```diff
@@ -0,0 +1,124 @@
+                return;
+            }
+
+            if (entBoardContainer.ContainedEntities.Count != 1)
+            {
+                Logger.Warning($"Machine frame entity {entity} did not have exactly one item in the '{MachineFrameComponent.BoardContainer}' container! Aborting build machine action.");
+            }
+
+            var board = entBoardContainer.ContainedEntities[0];
+
+            if (!board.TryGetComponent(out MachineBoardComponent? boardComponent))
+            {
+                Logger.Warning($"Machine frame entity {entity} had an invalid entity in container \"{MachineFrameComponent.BoardContainer}\"! Aborting build machine action.");
+                return;
+            }
+
+            var entityManager = entity.EntityManager;
+            entBoardContainer.Remove(board);
+
+            var machine = entityManager.SpawnEntity(boardComponent.Prototype, entity.Transform.Coordinates);
+            machine.Transform.LocalRotation = entity.Transform.LocalRotation;
+
+            // Cleanup method for containers.
+            static void CleanContainer(IContainer cont)
+            {
+                foreach (var ent in cont.ContainedEntities.ToArray())
+                {
+                    cont.ForceRemove(ent);
+                    ent.Delete();
+                }
+            }
+
+            var boardContainer = ContainerManagerComponent.Ensure<Container>(MachineFrameComponent.BoardContainer, machine, out var existed);
+
+            if (existed)
+            {
+                // Clean that up...
+                CleanContainer(boardContainer);
+            }
+
+            var partContainer = ContainerManagerComponent.Ensure<Container>(MachineFrameComponent.PartContainer, machine, out existed);
+
+            if (existed)
+            {
+                // Clean that up, too...
+                CleanContainer(partContainer);
+            }
+
+            boardContainer.Insert(board);
+
+            // Now we insert all parts.
+            foreach (var part in entPartContainer.ContainedEntities.ToArray())
+            {
+                entPartContainer.ForceRemove(part);
+                partContainer.Insert(part);
+            }
+
+            if (machine.TryGetComponent(out ConstructionComponent? construction))
+            {
+                // We only add these two container. If some construction needs to take other containers into account, fix this.
+                construction.AddContainer(MachineFrameComponent.BoardContainer);
+                construction.AddContainer(MachineFrameComponent.PartContainer);
+            }
+
+            if (machine.TryGetComponent(out MachineComponent? machineComp))
+            {
+                machineComp.RefreshParts();
+            }
+
+            entity.Delete();
+        }
+
+        public void ExposeData(ObjectSerializer serializer)
+        {
+        }
```

Comment:
No, but `IGraphAction` implements `IExposeData` since some actions can actually take serializable data and I need to implement these even if empty, sadly :(